src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -372,7 +373,12 @@ InterpreterSelectQuery::InterpreterSelectQuery(
         table_lock = storage->lockForShare(context->getInitialQueryId(), context->getSettingsRef().lock_acquire_timeout);
         table_id = storage->getStorageID();
         if (!metadata_snapshot)
-            metadata_snapshot = storage->getInMemoryMetadataPtr();
+        {
+            if (auto * window_view = dynamic_cast<StorageWindowView *>(storage.get()))
+                metadata_snapshot = window_view->getTargetTable()->getInMemoryMetadataPtr();
```

Comment:
because the structure of the columns is different when writing into windowview(getSampleBlock with stage FetchColumns, marks as `source columns` in following) and reading from windowview(getSampleBlock with stage Complete, marks as `target columns` in following). So we have to make a separate case for window view in reading or writing query.

In this PR, we change the columns in `InMemoryMetadataPtr` from `target columns` to `source columns` so that we can get the structure of the columns when windowview is attached and do not need to wait for the source table to attach(or sometimes source table might be dropped so we should not rely on the source table when attaching).

So we modify the interpreter of the select query and tell the interpreter the structure of the columns should be like the target table instead of the source table.

By the way, are we considering having something like `getWrittingColumns` and `getReadingColumns` for views that need to process the query inside the view (materialized view will process the query before insert)? Just like `getInputHeader` and `getOutputHeader` in `Chain`.

Reply:
> So we modify the interpreter of the select query and tell the interpreter the structure of the columns should be like the target table instead of the source table.

but then target table needs to be attached (cannot be known in case it is not inner table).

> By the way, are we considering having something like getWrittingColumns and getReadingColumns for views that need to process the query inside the view (materialized view will process the query before insert)? 

I think eventually yes (because looks like there is no other way to get rid of these special cases like this one for window view, which does not look right), need to think about it.

Reply:
> but then target table needs to be attached (cannot be known in case it is not inner table).

Yes, but we can assume that the target table is attached when executing a select query, otherwise, the select query is meaningless. Select from window view when target table not attached will throw exception `Table 'target' doesn't exist`, same as the materialized view.